src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -652,17 +684,17 @@ private Task ProcessDataFrameAsync(in ReadOnlySequence<byte> payload)
     {
         if (_currentHeadersStream != null)
         {
-            throw new Http2ConnectionErrorException(CoreStrings.FormatHttp2ErrorHeadersInterleaved(_incomingFrame.Type, _incomingFrame.StreamId, _currentHeadersStream.StreamId), Http2ErrorCode.PROTOCOL_ERROR);
+            throw CreateHeadersInterleavedException();
```

Comment:
Using a throw-helper would be IMO better here.
```c#
if (_currentHeadersStream != null)
{
    ThrowCreateHeadersInterleaved();
}

// ....

// Note: no inlining attributes here!
private void ThrowCreateHeadersInterleaved()
{
    // ...
}
```
So the JIT sees it's a non-returning method and won't inline it and will instead place it on a cold-path. Thus making the happy-path shorter.

The same on the other places.

PS: this could be done in a separate PR (I can look into this later on).

Reply:
Sure, doing it after this is merged would be good.